cql3/expression: implement evaluate(field_selection)

cvybhu opened the pull request:
Implement `expr:valuate()` for `expr::field_selection`.

`field_selection` is used to represent access to a struct field.
For example, with a UDT value:
```
CREATE TYPE my_type (a int, b int);
```
The expression `my_type_value.a` would be represented as a `field_selection`, which selects the field `a`.

Evaluating such an expression consists of finding the right element's value in a serialized UDT value and returning it.

Note that it's still not possible to use `field_selection` inside the `WHERE` clause. Enabling it would require changes to the grammar, as well as query planning, Current `statement_restrictions` just reacts with `on_internal_error` when it encounters a `field_selection`.
Nonetheless it's a step towards relaxing the grammar, and now it's finally possible to evaluate all kinds of prepared expressions (#12906)

Fixes: https://github.com/scylladb/scylladb/issues/12906

nyh commented:
Can you please say a few words how this fix relates (or doesn't relate) to #13624, which is about field subscripts not working in LWT? I thought it was only a grammar issue but maybe our expression evaluation also had missing pieces? Will this patch help but not completely fix #13624 (because the grammar parts are still missing), or it's unrelated?

cvybhu commented:
> Can you please say a few words how this fix relates (or doesn't relate) to https://github.com/scylladb/scylladb/issues/13624, which is about field subscripts not working in LWT? I thought it was only a grammar issue but maybe our expression evaluation also had missing pieces? Will this patch help but not completely fix https://github.com/scylladb/scylladb/issues/13624 (because the grammar parts are still missing), or it's unrelated?

Added a comment under the issue, so that it survives after this PR gets merged: https://github.com/scylladb/scylladb/issues/13624#issuecomment-1590574501

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/1806/

nyh commented on cql3/expr/expression.cc:
```diff
@@ -1659,6 +1659,53 @@ cql3::raw_value evaluate(const conjunction& conj, const evaluation_inputs& input
+    std::optional<std::size_t> selected_field_idx = udt_type->idx_of_field(to_bytes(selected_field_name));
+    if (!selected_field_idx.has_value()) {
+        throw exceptions::invalid_request_exception(
+            format("Unkown field '{}' in expression {}, user type {} doesn't have a field with this name.",
```
Typo, should be "Unknown".

cvybhu commented on cql3/expr/expression.cc:
```diff
@@ -1659,6 +1659,53 @@ cql3::raw_value evaluate(const conjunction& conj, const evaluation_inputs& input
+    // so we can use read_nth_tuple_elem to read the nth UDT field.
+    std::optional<managed_bytes_opt> element_value = udt_value.view().with_value(
+        [&](const FragmentedView auto& value_bytes) -> std::optional<managed_bytes_opt> {
+            // std::optional<std::optional<FragmentedView>> field_read
```
I didn't know how to write it when there's no clear way to get the type of `value_bytes`.
It's `const FragmentedView auto& value_bytes`, but how do I get what this `auto` is? `FragmentedView` is just a concept, not a type name.
I guess it's possible to acquire this type using some cool C++ code, but it was the quickest to just spell it out in the comment.

nyh replied:
I'm not familiar with "with_value()", doesn't it return a known type?
But if it doesn't, maybe something like
```
using value_type = std::remove_reference<decltype(value_bytes)>::type;
std::optional<std::optional<value_type>> ...
```
But that's not beautiful either... So you can just keep your version.

cvybhu replied:
> I'm not familiar with "with_value()", doesn't it return a known type?

`with_value` is a method that takes a closure which can accept argument of any type which fulfills the concept `FragmentedView`, and runs it on the `raw_value`. The argument is some view of the bytes, but we don't know which one, it could be fragmented or not.
The return type is known, but `value_bytes` has the `FragmentedView` type, which is different from the result type.
```c++
    template <typename Func>
    requires std::invocable<Func, const managed_bytes_view&> && std::invocable<Func, const fragmented_temporary_buffer::view&>
    decltype(auto) with_value(Func f) const {
        switch (_data.index()) {
        case 0: return f(std::get<fragmented_temporary_buffer::view>(_data));
        default: return f(std::get<managed_bytes_view>(_data));
        }
    }
```

cvybhu commented on types/types.hh:
```diff
@@ -962,6 +962,16 @@ utils::chunked_vector<managed_bytes_opt> partially_deserialize_listlike(View in)
+// The second std::nullopt signifies whether the value is NULL or not.
+// make_optional(make_optional(some bytes...)) - a non-null value with these bytes
+// make_optional(std::nullopt) - a null value
+// std::nullopt - no element with this index
```
I tried to make `read_nth_tuple_element` a generic function for getting a field of any tuple, so it made sense to notify the caller when any of those cases happen.

In UDTs the value is `null` in both cases (`null field`, `field not present`), but in normal tuples there is a difference. Maybe in the future we would like to be able to access any element of a tuple (`tuple_val[0]`/`tuple_val.0`). Now there's a difference between an existing `null` value and reading an out-of-bounds tuple element. I want this function to be reusable for getting tuple elements, not only UDTs.

I could add a separate function for getting the nth element of a UDT value, but I'm not sure if it's worth it.
It's important to know that for UDTs an out-of-bounds field is considered to have a NULL value, and people who look at/write the code which handles serialized UDTs should be aware of this, which can be forced by making them handle both cases explicitly.

nyh replied:
Doesn't your last paragraph actually "prove" that we need a read_udt_element() function that knows that an out of bounds field is considered NULL and not something else (not an error, etc.) - i.e., that this logic needs to be written once in the function instead of every caller needing to remember how to handle the two levels of nullopt separately?

I personally think (but don't insist if you think otherwise - you're more an expert in this than me) that if the UDT indexing feature has UDT-specific rules, these rules should be written in a separate UDT-specific function instead of all callers remembering that a UDT is really a tuple with special rules. Moreover, where can the function you added, `read_nth_tuple_element` be used in the future on real tuples? If I understand correctly, CQL doesn't have a syntax to index tuples. Or were you thinking of may adding such a syntax? In that case, wouldn't an out-of-bound index also want to return a null and not anything else?

cvybhu replied:
Ok, added `read_nth_user_type_field`

cvybhu commented on cql3/expr/expression.cc:
```diff
@@ -1659,6 +1659,53 @@ cql3::raw_value evaluate(const conjunction& conj, const evaluation_inputs& input
+    if (!selected_field_idx.has_value()) {
+        throw exceptions::invalid_request_exception(
+            format("Unkown field '{}' in expression {}, user type {} doesn't have a field with this name.",
+                   selected_field_name, field_select, udt_type->get_name_as_cql_string()));
```
:+1: removed the copy

cvybhu commented on cql3/expr/expression.cc:
```diff
@@ -1659,6 +1659,53 @@ cql3::raw_value evaluate(const conjunction& conj, const evaluation_inputs& input
+    }
+
+    sstring selected_field_name = field_select.field->to_string();
+    std::optional<std::size_t> selected_field_idx = udt_type->idx_of_field(to_bytes(selected_field_name));
```
:+1: removed the copy

cvybhu commented:
v2:
* Added `read_nth_user_type_field()` in `types.hh`, this made `evaluate()` easier to read.
* Removed string copies from `evaluate()`, needed to modify the APIs a bit
* Fixed `Unkown` typo
* Using `get_name_as_string` in the error message

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/1871/
